### PR TITLE
Feature authority transfer

### PR DESCRIPTION
**Pull request by agrinman**

### CODE REVIEW

File: protocol.go [hunk 1]
```diff
@@ -108,6 +118,22 @@ type GitSignResponse struct {
 	Error     *string `json:"error,omitempty"`
 }
 
+type HostsRequest struct{}
+
+type UserAndHost struct {
+	User string `json:"user,omitempty"`
```
Comment (kcking) [inline protocol.go]: omitempty should only be used on pointer members

File: kr/kr.go [hunk 2]
```diff
@@ -628,6 +658,17 @@ func main() {
 			},
 		},
 		cli.Command{
+			Name:   "transfer",
+			Usage:  "Transfer authority to a new Kryptonite public key (authorize a new Kryptonite device's public key to servers)",
+			Action: transferCommand,
+			Flags: []cli.Flag{
+				cli.BoolFlag{
+					Name:  "dry-run, d",
+					Usage: "Do a dry-run and preview all servers that kr will try to authorize the new public key too.",
```
Comment (kcking) [inline kr/kr.go]: s/authorize/add

File: kr/kr_darwin.go [hunk 3]
```diff
@@ -110,7 +110,7 @@ func copyEnvToLaunchctl(varName string) {
 	_, _ = runCommandTmuxFriendly("launchctl", "setenv", varName, os.Getenv(varName))
 }
 
-func restartCommand(c *cli.Context) (err error) {
+func restartCommandOptions(c *cli.Context, showPrints bool) (err error) {
```
Comment (kcking) [inline kr/kr_darwin.go]: with this refactor the analytics will print even for a transfer command. can we rename `showPrints` and only have send analytics when it is true?

File: kr/kr.go [hunk 4]
```diff
@@ -84,6 +84,28 @@ func pairCommand(c *cli.Context) (err error) {
 	return pairOver(kr.DaemonSocketOrFatal(), c.Bool("force"), nameOpt, os.Stdout, os.Stderr)
 }
 
+func pairCommandForce() (err error) {
```
Comment (kcking) [inline kr/kr.go]: can we prevent analytics when being invoked by kr transfer? if this is only used in kr transfer then analytics can just be deleted

File: kr/kr_unix.go [hunk 5]
```diff
@@ -20,11 +20,14 @@ func getPrefix() string {
...
 }
 
-func restartCommand(c *cli.Context) (err error) {
+func restartCommandOptions(c *cli.Context, showPrints bool) (err error) {
 	kr.Analytics{}.PostEventUsingPersistedTrackingID("kr", "restart", nil, nil)
 	exec.Command("pkill", "krd").Run()
 	startKrd()
-	PrintErr(os.Stderr, "Restarted Kryptonite daemon.")
+
+	if showPrints {
```
Comment (kcking) [inline kr/kr_unix.go]: same here regarding `showPrints` renaming and analytics

File: kr/kr.go (hunk 4 above)
Comment (agrinman) [reply kr/kr.go]: only in transfer, will remove analytics line 👍

File: kr/kr.go (hunk 2 above)
Comment (agrinman) [reply kr/kr.go]: 👍

File: kr/kr_darwin.go (hunk 3 above)
Comment (agrinman) [reply kr/kr_darwin.go]: `showPrints` -> `isUserInitiated` 👍

File: kr/kr_unix.go (hunk 5 above)
Comment (agrinman) [reply kr/kr_unix.go]: 👍

File: kr/transfer_authority.go [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+
+func getFilePersister() (files kr.FilePersister, err error) {
+	krdir, err := kr.KrDir()
+	if err != nil {
+		return
+	}
+
+	files = kr.FilePersister{
+		PairingDir: krdir,
+		SSHDir:     filepath.Join(kr.UnsudoedHomeDir(), ".ssh")}
```
[the remark this answers is not in the document]
Comment (agrinman) [reply kr/transfer_authority.go]: 👍

File: kr/transfer_authority.go [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+
+	return
+}
+
+func doRequestHostInfo() (hostInfo kr.HostInfo, err error) {
+	os.Stderr.WriteString(kr.Cyan("Kryptonite ▶ Requesting logs from phone") + "\r\n")
+	// os.Stderr.WriteString(kr.Yellow("Kryptonite ▶ Phone approval required. Response using the Kryptonite app") + "\r\n")
```
[the remark this answers is not in the document]
Comment (agrinman) [reply kr/transfer_authority.go]: 👍

File: kr/transfer_authority.go [hunk 8]
```diff
@@ -0,0 +1,490 @@
...
+	}
+	os.Stderr.WriteString("\n" + kr.Green("Success, paired with NEW Kryptonite device ✔") + "\r\n")
+
+	newAuthorizedPublicKey, err := newProfile.AuthorizedKeyString()
+	if err != nil {
+		return
+	}
+
+	/// manually re-pair with old session
+	doManualRePairWithOldKryptonite(c, oldProfile)
```
[the remark this answers is not in the document]
Comment (agrinman) [reply kr/transfer_authority.go]: 👍

File: kr/transfer_authority.go [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+			os.Stderr.WriteString(kr.Red("× failed authorizing "+host.User+" @ "+host.Host) + "\r\n")
+			continue
+		}
+	}
+
+	// show failed hosts
+	handleFailures(failedHosts)
+
+	/// manually re-pair with new session
+	doManualRePairWithNewKryptonite(c, newProfile)
```
[the remark this answers is not in the document]
Comment (agrinman) [reply kr/transfer_authority.go]: 👍

File: protocol.go (hunk 1 above)
Comment (agrinman) [reply protocol.go]: right, 👍

**Pull request merged**